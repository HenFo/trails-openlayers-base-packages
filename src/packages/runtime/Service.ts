export interface ServiceLifecycleHooks {
    /**
     * Destroy the component.
     * Should clean up all resources used by the component.
     */
    destroy?(): void;

    /*
     * Just used so that this interface is not empty.
     * TypeScript would otherwise throw an 'no common properties' error
     * if one tries to implement this interface without providing any
     * lifecycle hooks.
     */
    toString(): string;
}

export type Service<Interface extends {} = {}> = ServiceLifecycleHooks & Interface;

export type ServiceOptions<References extends {} = {}> = {
    /**
     * References to other services.
     * These are injected by the runtime and match the declared references of the service.
     */
    references: References;

    /**
     * Configuration properties.
     *
     * TODO: These are not specified yet
     */
    properties: Record<string, unknown>;
};

export type ServiceConstructor<References extends {} = {}, Interface extends {} = {}> = {
    new (options: ServiceOptions<References>): Service<Interface>;
};
