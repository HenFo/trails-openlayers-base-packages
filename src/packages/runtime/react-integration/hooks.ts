import { useContext, useMemo } from "react";
import { Error } from "@open-pioneer/core";
import { InterfaceName, ServiceType } from "../ServiceRegistry";
import { PackageContext as PackageContext, PackageContextData } from "./PackageContext";
import { ErrorId } from "../errors";

/**
 * Returns a service that implements the given interface.
 * Receives the package name of the importing package as a string.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useServiceInternal<Interface extends InterfaceName>(
    packageName: string,
    interfaceName: Interface
): ServiceType<Interface>;
export function useServiceInternal(packageName: string, interfaceName: string): unknown;
export function useServiceInternal(packageName: string, interfaceName: string) {
    const context = useContext(PackageContext);
    const service = useMemo(
        () => checkContext("useService", context).getService(packageName, interfaceName),
        [context, packageName, interfaceName]
    );
    return service;
}

/**
 * Returns the properties for the given package.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function usePropertiesInternal(packageName: string): Readonly<Record<string, unknown>> {
    const context = useContext(PackageContext);
    return checkContext("useProperties", context).getProperties(packageName);
}

function checkContext(
    hookName: string,
    contextData: PackageContextData | null
): PackageContextData {
    if (!contextData) {
        throw new Error(
            ErrorId.INTERNAL,
            `"Failed to access package context from '${hookName}': react integration was not set up properly.`
        );
    }
    return contextData;
}
