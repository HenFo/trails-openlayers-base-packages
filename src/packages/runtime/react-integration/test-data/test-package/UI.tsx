import { useProperties } from "open-pioneer:react-hooks";

export function UI() {
    const properties = useProperties();
    return <div className="ui">{properties.greeting as string}</div>
}
