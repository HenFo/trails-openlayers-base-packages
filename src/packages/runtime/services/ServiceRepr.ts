import { ErrorId } from "./../errors";
import { InterfaceReferenceMetadata, ServiceMetadata } from "../Metadata";
import { Service, ServiceConstructor, ServiceOptions } from "../Service";
import { Error } from "@open-pioneer/core";

export type Dependency = { name: string } & InterfaceReferenceMetadata;

export type ServiceState = "not-constructed" | "constructing" | "constructed" | "destroyed";

/**
 * Represents metadata and state of a service in the runtime.
 * `this.instance` is the actual service instance (when constructed).
 */
export class ServiceRepr {
    static parse(bundleName: string, data: ServiceMetadata): ServiceRepr {
        const clazz = data.clazz;
        const name = data.name;
        const dependencies = Object.entries(data.references ?? {}).map<Dependency>(
            ([name, referenceMetadata]) => {
                return {
                    name,
                    ...referenceMetadata
                };
            }
        );
        // TODO: Properties in metadata?
        const interfaces = (data.provides ?? []).map((p) => p.interface);
        return new ServiceRepr(name, bundleName, clazz, dependencies, interfaces, {});
    }

    /** Unique id of this service. Contains the bundle name and the service name. */
    readonly id: string;

    /** Name of this service in it's bundle. */
    readonly name: string;

    /** Name of the parent bundle. */
    readonly bundleName: string;

    /** Service properties made available via the service's constructor. */
    readonly properties: Readonly<Record<string, unknown>>;

    /** Dependencies required by the service constructor. */
    readonly dependencies: readonly Dependency[];

    /** Interfaces provided by the service. */
    readonly interfaces: readonly string[];

    /** Number of references to this service. */
    private useCount = 0;

    /** Service constructor. */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private clazz: ServiceConstructor<any>;

    /** Current state of this service. "constructed" -> instance is available. */
    private _state: ServiceState = "not-constructed";

    /** Service instance, once constructed. */
    private _instance: Service | undefined = undefined;

    constructor(
        name: string,
        bundleName: string,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        clazz: ServiceConstructor<any>,
        dependencies: Dependency[],
        interfaces: string[],
        properties: Record<string, unknown>
    ) {
        this.id = `${bundleName}::${name}`;
        this.name = name;
        this.bundleName = bundleName;
        this.clazz = clazz;
        this.dependencies = dependencies;
        this.interfaces = interfaces;
        this.properties = properties;
    }

    /** Returns the current service instance or undefined if the service has not been constructed. */
    get instance(): Service | undefined {
        return this._instance;
    }

    /** Returns the current state of the service. */
    get state() {
        return this._state;
    }

    /**
     * Same as `instance`, but throws when the instance has not been constructed.
     */
    getInstanceOrThrow() {
        const instance = this._instance;
        if (!instance) {
            throw new Error(ErrorId.INTERNAL, "Expected service instance to be present.");
        }
        return instance;
    }

    beforeCreate() {
        if (this._state === "not-constructed") {
            this._state = "constructing";
        } else {
            throw new Error(
                ErrorId.INTERNAL,
                "Inconsistent state: service is already under construction."
            );
        }
    }

    /**
     * Instantiates the service by invoking the service constructor 
     * with the given `options`.
     * 
     * The service's use count is initialized to `1`, so every `create()`
     * should be paired with a `removeRef()`.
     * 
     * `destroy()` can be invoked once the final `removeRef()` has returned zero. 
     */
    create(options: ServiceOptions) {
        if (this._state !== "constructing" || this.instance !== undefined) {
            throw new Error(
                ErrorId.INTERNAL,
                "Inconsistent state: service is not being constructed."
            );
        }
        try {
            this._instance = new this.clazz(options);
            this._state = "constructed";
            this.useCount = 1;
            return this._instance;
        } catch (e) {
            throw new Error(
                ErrorId.SERVICE_CONSTRUCTION_FAILED,
                `Failed to create service '${this.id}'.`,
                { cause: e }
            );
        }
    }

    destroy() {
        if (this._instance) {
            try {
                this._instance.destroy?.();
            } catch (e) {
                throw new Error(
                    ErrorId.SERVICE_DESTRUCTION_FAILED,
                    `Failed to destroy service '${this.id}'.`,
                    { cause: e }
                );
            }
        }
        this._instance = undefined;
        this._state = "destroyed";
    }

    /**
     * Adds a use to the service's use count.
     * References to a service are tracked: it should only be destroyed when it is no longer being used.
     */
    addRef() {
        return this.useCount += 1;
    }

    /**
     * Removes a use from the service's use count.
     * Returns the new use count.
     */
    removeRef() {
        return this.useCount -= 1;
    }
}
