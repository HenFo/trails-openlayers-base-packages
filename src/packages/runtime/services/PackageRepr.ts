import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { PackageMetadata } from "../Metadata";
import { ServiceRepr } from "./ServiceRepr";

export class PackageRepr {
    static parse(data: PackageMetadata): PackageRepr {
        const name = data.name;
        const services = Object.entries(data.services).map<ServiceRepr>(([name, serviceData]) => {
            if (name !== serviceData.name) {
                throw new Error(
                    ErrorId.INVALID_METADATA,
                    "Invalid metadata: service name mismatch."
                );
            }
            return ServiceRepr.parse(data.name, serviceData);
        });
        return new PackageRepr(name, services);
    }

    readonly name: string;
    readonly services: readonly ServiceRepr[];

    constructor(name: string, services: ServiceRepr[]) {
        this.name = name;
        this.services = services;
    }
}

export function parsePackages(packages: Record<string, PackageMetadata>) {
    return Object.entries(packages).map<PackageRepr>(([name, packageMetadata]) => {
        if (name !== packageMetadata.name) {
            throw new Error(ErrorId.INVALID_METADATA, "Invalid metadata: package name mismatch.");
        }

        return PackageRepr.parse(packageMetadata);
    });
}
