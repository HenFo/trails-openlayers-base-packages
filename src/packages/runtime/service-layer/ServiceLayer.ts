import { ServiceRepr } from "./ServiceRepr";
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import {
    ComputedServiceDependencies,
    UIDependency,
    verifyDependencies
} from "./verifyDependencies";
import { PackageRepr } from "./PackageRepr";
import { ReadonlyServiceLookup, ServiceLookupResult, ServicesLookupResult } from "./ServiceLookup";
import {
    InterfaceSpec,
    isAllImplementationsSpec,
    isSingleImplementationSpec,
    ReferenceSpec
} from "./InterfaceSpec";
import { ReferenceMeta } from "../Service";

export type DynamicLookupResult = ServiceLookupResult | UndeclaredDependency;

export interface UndeclaredDependency {
    type: "undeclared";
}

interface DependencyDeclarations {
    all: boolean;
    unqualified: boolean;
    qualifiers: Set<string>;
}

export class ServiceLayer {
    private serviceLookup: ReadonlyServiceLookup;

    // Service id --> Service dependencies
    private serviceDependencies: Map<string, ComputedServiceDependencies>;

    // Package name --> Interface name --> Declarations
    private declaredDependencies;
    private allServices: readonly ServiceRepr[];
    private state: "not-started" | "started" | "destroyed" = "not-started";

    constructor(packages: readonly PackageRepr[]) {
        this.allServices = packages.map((pkg) => pkg.services).flat();
        const { serviceLookup, serviceDependencies } = verifyDependencies({
            services: this.allServices,
            uiDependencies: packages
                .map((pkg) =>
                    pkg.uiReferences.map<UIDependency>((dep) => {
                        return {
                            packageName: pkg.name,
                            ...dep
                        };
                    })
                )
                .flat()
        });
        this.serviceLookup = serviceLookup;
        this.serviceDependencies = serviceDependencies;
        this.declaredDependencies = buildDependencyIndex(packages);
    }

    destroy() {
        this.allServices.forEach((value) => {
            this.destroyService(value);
        });
        this.state = "destroyed";
    }

    start() {
        if (this.state !== "not-started") {
            throw new Error(ErrorId.INTERNAL, "Service layer was already started.");
        }

        this.allServices.forEach((value) => {
            this.createService(value);
        });
        this.state = "started";
    }

    /**
     * Returns a service implementing the given interface.
     * Checks that the given package actually declared a dependency on that interface
     * to enforce coding guidelines.
     *
     * @param packageName the name of the package requesting the import
     * @param spec the interface specifier
     * @param options advanced options to customize lookup
     * @throws if the service layer is not in 'started' state or if no service implements the interface.
     */
    getService(
        packageName: string,
        spec: InterfaceSpec,
        options?: { ignoreDeclarationCheck?: boolean }
    ): ServiceLookupResult | UndeclaredDependency {
        if (this.state !== "started") {
            throw new Error(ErrorId.INTERNAL, "Service layer is not started.");
        }

        if (!options?.ignoreDeclarationCheck && !this.isDeclaredDependency(packageName, spec)) {
            return { type: "undeclared" };
        }

        return this.serviceLookup.lookup(spec);
    }

    /**
     * Returns all services implementing the given interface.
     * Requires that the dependency on all implementations has been declared in the package.
     *
     * @param packageName the name of the package requesting the import
     * @param interfaceName the interface name
     *
     * @throws if the service layer is not in 'started'.
     */
    getServices(
        packageName: string,
        interfaceName: string
    ): ServicesLookupResult | UndeclaredDependency {
        if (this.state !== "started") {
            throw new Error(ErrorId.INTERNAL, "Service layer is not started.");
        }

        if (!this.isDeclaredDependency(packageName, { interfaceName, all: true })) {
            return { type: "undeclared" };
        }

        return this.serviceLookup.lookupAll(interfaceName);
    }

    /**
     * Initializes the given service and its dependencies.
     * Dependencies are initialized before the service that requires them.
     */
    private createService(service: ServiceRepr) {
        if (service.state === "constructed") {
            const instance = service.getInstanceOrThrow();
            service.addRef();
            return instance;
        }
        if (service.state === "constructing") {
            throw new Error(ErrorId.INTERNAL, "Cycle during service construction.");
        }
        if (service.state !== "not-constructed") {
            throw new Error(ErrorId.INTERNAL, "Invalid service state.");
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const references: Record<string, any> = {};
        const referencesMeta: Record<string, ReferenceMeta | ReferenceMeta[]> = {};

        // Sets state to 'constructing' to detect cycles
        service.beforeCreate();

        // Initialize dependencies recursively before creating the current service.
        for (const [referenceName, serviceDeps] of Object.entries(this.getServiceDeps(service))) {
            const [referenceValue, referenceMeta] = this.getReference(serviceDeps);
            references[referenceName] = referenceValue;
            referencesMeta[referenceName] = referenceMeta;
        }

        // Sets state to 'constructed' to finish the state transition, useCount is 1.
        return service.create({ references, referencesMeta, properties: service.properties });
    }

    /**
     * Destroys the given service and its dependencies.
     * The dependencies are destroyed after the service.
     */
    private destroyService(service: ServiceRepr) {
        if (service.state === "destroyed") {
            return;
        }

        // Destroy the service before its dependencies (reverse order
        // compared to construction).
        if (service.removeRef() <= 0) {
            service.destroy();
        }

        for (const serviceDeps of Object.values(this.getServiceDeps(service))) {
            if (Array.isArray(serviceDeps)) {
                for (const dep of serviceDeps) {
                    this.destroyService(dep);
                }
            } else {
                this.destroyService(serviceDeps);
            }
        }
    }

    private isDeclaredDependency(packageName: string, spec: ReferenceSpec) {
        const packageEntry = this.declaredDependencies.get(packageName);
        if (!packageEntry) {
            return false;
        }
        const interfaceEntry = packageEntry.get(spec.interfaceName);
        if (!interfaceEntry) {
            return false;
        }

        if (isSingleImplementationSpec(spec)) {
            if (spec.qualifier == null) {
                return interfaceEntry.unqualified;
            }
            return interfaceEntry.qualifiers.has(spec.qualifier);
        } else {
            return interfaceEntry.all;
        }
    }

    private getReference(ref: ServiceRepr): [unknown, ReferenceMeta];
    private getReference(ref: ServiceRepr[]): [unknown[], ReferenceMeta[]];
    private getReference(
        ref: ServiceRepr | ServiceRepr[]
    ): [unknown, ReferenceMeta] | [unknown[], ReferenceMeta[]];
    private getReference(
        ref: ServiceRepr | ServiceRepr[]
    ): [unknown, ReferenceMeta] | [unknown[], ReferenceMeta[]] {
        if (Array.isArray(ref)) {
            const referenceValues: unknown[] = [];
            const referenceMeta: ReferenceMeta[] = [];
            for (const d of ref) {
                const [value, meta] = this.getReference(d);
                referenceValues.push(value);
                referenceMeta.push(meta);
            }
            return [referenceValues, referenceMeta];
        }

        const referenceValue = this.createService(ref);
        const referenceMeta: ReferenceMeta = {
            serviceId: ref.id
        };
        return [referenceValue, referenceMeta];
    }

    private getServiceDeps(service: ServiceRepr) {
        const dependencies = this.serviceDependencies.get(service.id);
        if (!dependencies) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to find precomputed service dependencies for '${service.id}'.`
            );
        }
        return dependencies;
    }
}

function buildDependencyIndex(packages: readonly PackageRepr[]) {
    // Register declared UI dependencies.
    // This is needed as a lookup structure to that dynamic service lookups
    // can be validated at runtime.
    const index = new Map<string, Map<string, DependencyDeclarations>>();
    for (const pkg of packages) {
        const packageName = pkg.name;
        const packageEntry = new Map<string, DependencyDeclarations>();
        for (const uiReference of pkg.uiReferences) {
            let interfaceEntry = packageEntry.get(uiReference.interfaceName);
            if (!interfaceEntry) {
                interfaceEntry = {
                    all: false,
                    unqualified: false,
                    qualifiers: new Set<string>()
                };
                packageEntry.set(uiReference.interfaceName, interfaceEntry);
            }

            if (isAllImplementationsSpec(uiReference)) {
                interfaceEntry.all = true;
            } else if (uiReference.qualifier == null) {
                interfaceEntry.unqualified = true;
            } else {
                interfaceEntry.qualifiers.add(uiReference.qualifier);
            }
        }
        index.set(packageName, packageEntry);
    }
    return index;
}
