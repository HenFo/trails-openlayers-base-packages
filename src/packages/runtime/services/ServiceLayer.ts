import { ServiceRepr } from "./ServiceRepr";
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { UIDependency, verifyDependencies } from "./verifyDependencies";
import { PackageRepr } from "./PackageRepr";
import { Service } from "../Service";

export type ServiceLookupResult = FoundService | UnimplementedService | UndeclaredDependency;

export interface FoundService {
    type: "found";
    instance: Service;
}

export interface UndeclaredDependency {
    type: "undeclared";
}

export interface UnimplementedService {
    type: "unimplemented";
}

export class ServiceLayer {
    readonly serviceIndex: ReadonlyMap<string, ServiceRepr>;

    private declaredDependencies = new Map<string, Set<string>>();
    private allServices: readonly ServiceRepr[];
    private state: "not-started" | "started" | "destroyed" = "not-started";

    constructor(packages: readonly PackageRepr[]) {
        this.allServices = packages.map((pkg) => pkg.services).flat();
        this.serviceIndex = verifyDependencies({
            services: this.allServices,
            uiDependencies: packages
                .map((pkg) =>
                    pkg.uiInterfaces.map<UIDependency>((interfaceName) => {
                        return {
                            packageName: pkg.name,
                            interfaceName
                        };
                    })
                )
                .flat()
        });

        for (const pkg of packages) {
            this.declaredDependencies.set(pkg.name, new Set(pkg.uiInterfaces));
        }
    }

    destroy() {
        this.allServices.forEach((value) => {
            this.destroyService(value);
        });
        this.state = "destroyed";
    }

    start() {
        if (this.state !== "not-started") {
            throw new Error(ErrorId.INTERNAL, "Service layer was already started.");
        }

        this.allServices.forEach((value) => {
            this.createService(value);
        });
        this.state = "started";
    }

    /**
     * Returns a service implementing the given interface.
     * Checks that the given package actually declared a dependency on that interface
     * to enforce coding guidelines.
     *
     * @param packageName the name of the package requesting the import
     * @param interfaceName the required interface
     *
     * @throws if the service layer is not in 'started' state or if no service implements the interface.
     */
    getService(packageName: string, interfaceName: string): ServiceLookupResult {
        if (this.state !== "started") {
            throw new Error(ErrorId.INTERNAL, "Service layer is not started.");
        }

        const instance = this.serviceIndex.get(interfaceName)?.instance;
        if (!instance) {
            return { type: "unimplemented" };
        }

        const isDeclared = this.declaredDependencies.get(packageName)?.has(interfaceName) ?? false;
        if (!isDeclared) {
            return { type: "undeclared" };
        }

        return {
            type: "found",
            instance: instance
        };
    }

    /**
     * Initializes the given service and its dependencies.
     * Dependencies are initialized before the service that requires them.
     */
    private createService(service: ServiceRepr) {
        if (service.state === "constructed") {
            const instance = service.getInstanceOrThrow();
            service.addRef();
            return instance;
        }
        if (service.state === "constructing") {
            throw new Error(ErrorId.INTERNAL, "Cycle during service construction.");
        }
        if (service.state !== "not-constructed") {
            throw new Error(ErrorId.INTERNAL, "Invalid service state.");
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const instances: Record<string, any> = {};

        // Sets state to 'constructing' to detect cycles
        service.beforeCreate();

        // Initialize dependencies recursively before creating the current service.
        service.dependencies.forEach((d) => {
            const serviceRef = this.serviceIndex.get(d.interfaceName);
            if (serviceRef) {
                const instance = this.createService(serviceRef);
                instances[d.referenceName] = instance;
            } else {
                throw new Error(ErrorId.INTERNAL, "Service not defined.");
            }
        });

        // Sets state to 'constructed' to finish the state transition, useCount is 1.
        return service.create({ references: instances, properties: service.properties });
    }

    /**
     * Destroys the given service and its dependencies.
     * The dependencies are destroyed after the service.
     */
    private destroyService(service: ServiceRepr) {
        if (service.state === "destroyed") {
            return;
        }

        // Destroy the service before its dependencies (reverse order
        // compared to construction).
        if (service.removeRef() <= 0) {
            service.destroy();
        }

        service.dependencies.forEach((d) => {
            const serviceRef = this.serviceIndex.get(d.interfaceName);
            if (serviceRef) {
                this.destroyService(serviceRef);
            }
        });
    }
}
