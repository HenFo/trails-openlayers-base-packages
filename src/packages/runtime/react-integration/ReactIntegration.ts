import { ComponentType, createElement, StrictMode } from "react";
import { createRoot, Root } from "react-dom/client";
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { ServiceLayer } from "../services/ServiceLayer";
import { PackageContext, PackageContextData } from "./PackageContext";
import { PackageRepr } from "../services/PackageRepr";

export interface ReactIntegrationOptions {
    packages: Map<string, PackageRepr>;
    serviceLayer: ServiceLayer;
    rootNode: HTMLDivElement;
}

export class ReactIntegration {
    private packages: Map<string, PackageRepr>;
    private serviceLayer: ServiceLayer;
    private root: Root;
    private packageContext: PackageContextData;

    constructor(options: ReactIntegrationOptions) {
        this.packages = options.packages;
        this.serviceLayer = options.serviceLayer;
        this.root = createRoot(options.rootNode);
        this.packageContext = {
            getService: (packageName, interfaceName) => {
                const result = this.serviceLayer.getService(packageName, interfaceName);
                if (result.type === "unimplemented") {
                    throw new Error(
                        ErrorId.INTERFACE_NOT_FOUND,
                        `The UI of package '${packageName}' requested the unimplemented interface '${interfaceName}'.`
                    );
                }
                if (result.type === "undeclared") {
                    throw new Error(
                        ErrorId.UNDECLARED_DEPENDENCY,
                        `Package '${packageName}' did not declare an UI dependency on interface '${interfaceName}'.` +
                            ` Add the dependency to the package configuration or remove the usage.`
                    );
                }
                return result.instance;
            },
            getProperties: (packageName) => {
                const pkg = this.packages.get(packageName);
                if (!pkg) {
                    throw new Error(
                        ErrorId.INTERNAL,
                        `Package '${packageName}' was not found in application.`
                    );
                }
                return pkg.properties;
            }
        };
    }

    render(component: ComponentType, props: Record<string, unknown>) {
        const element = createElement(component, props);
        const contextWrapper = createElement(
            PackageContext.Provider,
            { value: this.packageContext },
            element
        );
        this.root.render(createElement(StrictMode, undefined, contextWrapper));
    }

    destroy() {
        this.root.unmount();
    }
}
