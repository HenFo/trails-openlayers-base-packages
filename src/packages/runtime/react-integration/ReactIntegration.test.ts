/**
 * @vitest-environment jsdom
 */
import { createElement } from "react";
import { beforeEach, expect, it } from "vitest";
import { usePropertiesInternal, useServiceInternal } from "./hooks";
import { findByText } from "@testing-library/dom";
import { Service } from "../Service";
// eslint-disable-next-line import/no-relative-packages
import { UI as TestUIFromPackage } from "./test-data/test-package/UI";
import { ServiceLayer } from "../services/ServiceLayer";
import { ReactIntegration } from "./ReactIntegration";
import { act } from "react-dom/test-utils";
import { PackageRepr } from "../services/PackageRepr";
import { ServiceRepr } from "../services/ServiceRepr";

export interface TestProvider {
    value: string;
}

beforeEach(() => {
    document.body.innerHTML = "";
});

it("should allow access to service via react hook", async () => {
    function TestComponent() {
        const service = useServiceInternal("test", "test.Provider") as TestProvider;
        return createElement("span", undefined, `Hello ${service.value}`);
    }

    const wrapper = document.createElement("div");
    const testService = new ServiceRepr({
        name: "Provider",
        packageName: "test",
        clazz: class Provider implements Service<TestProvider> {
            value = "TEST";
        },
        interfaces: ["test.Provider"]
    });
    const testPackage = new PackageRepr({
        name: "test",
        uiInterfaces: ["test.Provider"],
        services: [testService]
    });
    const serviceLayer = new ServiceLayer([testPackage]);
    serviceLayer.start();

    const reactIntegration = new ReactIntegration({
        rootNode: wrapper,
        container: wrapper,
        serviceLayer,
        packages: new Map([[testPackage.name, testPackage]])
    });

    act(() => {
        reactIntegration.render(TestComponent, {});
    });

    const node = await findByText(wrapper, "Hello TEST");
    expect(node).toMatchSnapshot();
});

it("should be able to read properties from react component", async () => {
    function TestComponent() {
        const properties = usePropertiesInternal("test");
        return createElement("span", undefined, `Hello ${properties.name}`);
    }

    const wrapper = document.createElement("div");
    const serviceLayer = new ServiceLayer([]);
    serviceLayer.start();

    const testPackage = new PackageRepr({
        name: "test",
        properties: {
            name: "USER"
        }
    });
    const reactIntegration = new ReactIntegration({
        rootNode: wrapper,
        container: wrapper,
        serviceLayer,
        packages: new Map([[testPackage.name, testPackage]])
    });

    act(() => {
        reactIntegration.render(TestComponent, {});
    });

    const node = await findByText(wrapper, "Hello USER");
    expect(node).toMatchSnapshot();
});

it("should provide the autogenerated useProperties hook", async () => {
    const testPackageName = "@open-pioneer/runtime__react_test_package";
    const wrapper = document.createElement("div");
    const serviceLayer = new ServiceLayer([]);
    serviceLayer.start();

    const testPackage = new PackageRepr({
        name: testPackageName,
        properties: {
            greeting: "Hello World!"
        }
    });
    const reactIntegration = new ReactIntegration({
        rootNode: wrapper,
        container: wrapper,
        serviceLayer,
        packages: new Map([[testPackage.name, testPackage]])
    });

    act(() => {
        reactIntegration.render(TestUIFromPackage, {});
    });

    const node = await findByText(wrapper, "Hello World!");
    expect(node).toMatchSnapshot();
});

it("should get error when using undefined service", async () => {
    function TestComponent() {
        const service = useServiceInternal("test", "test.Provider") as TestProvider;
        return createElement("span", undefined, `Hello ${service.value}`);
    }
    const wrapper = document.createElement("div");
    const serviceLayer = new ServiceLayer([]);
    serviceLayer.start();

    const reactIntegration = new ReactIntegration({
        rootNode: wrapper,
        container: wrapper,
        serviceLayer,
        packages: new Map()
    });

    expect(() => {
        act(() => {
            reactIntegration.render(TestComponent, {});
        });
    }).toThrowErrorMatchingSnapshot();
});

it("should get error when requesting properties from an unknown package", async () => {
    const wrapper = document.createElement("div");
    const serviceLayer = new ServiceLayer([]);
    const reactIntegration = new ReactIntegration({
        rootNode: wrapper,
        container: wrapper,
        serviceLayer,
        packages: new Map()
    });

    function TestComponent() {
        const properties = usePropertiesInternal("test");
        return createElement("span", undefined, `Hello ${properties.name}`);
    }

    serviceLayer.start();

    expect(() => {
        act(() => {
            reactIntegration.render(TestComponent, {});
        });
    }).toThrowErrorMatchingSnapshot();
});
