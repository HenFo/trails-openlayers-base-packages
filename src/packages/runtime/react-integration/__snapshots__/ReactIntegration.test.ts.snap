// Vitest Snapshot v1

exports[`should allow access to service via react hook 1`] = `
<span>
  Hello TEST
</span>
`;

exports[`should be able to read properties from react component 1`] = `
<span>
  Hello USER
</span>
`;

exports[`should get error when requesting properties from an unknown package 1`] = `"runtime:internal-error: Package 'test' was not found in application."`;

exports[`should get error when using undefined service 1`] = `"runtime:interface-not-found: The UI of package 'test' requested the unimplemented interface 'test.Provider'."`;

exports[`should provide the autogenerated useProperties hook 1`] = `
<div
  class="ui"
>
  Hello World!
</div>
`;
